### PR TITLE
[feat][broker] ns-isolation-policy support hostport pattern

### LINKED ISSUE #15795: 在配置ns-isolation-policy 的时候broker的配置不能采用hostname:端口 但是
1  用命令行列出broker的时候
bin/pulsar-admin brokers list

“hostname1:8080”
“hostname2:8080”
“hostname3:8080”

bin/pulsar-admin   ns-isolation-policy  brokers
{
  "brokerName" : "hostname1:8080",
  "namespaceRegex" : [ ],
  "primary" : false
}
{
  "brokerName" : "hostname2:8080",
  "namespaceRegex" : [ ],
  "primary" : false
}
{
  "brokerName" : "hostname3:8080",
  "namespaceRegex" : [ ],
  "primary" : false
}

2 bin/pulsar-admin ns-isolation-policy set 省略其他参数 --primary hostname1:8080  --secondary  hostname2:8080
可以设置 但是无效

3 bin/pulsar-admin ns-isolation-policy set 省略其他参数 --primary hostname1  --secondary  hostname2
可以设置 有效

建议1：统一broker list的格式
5月份的开发组沟通会上确认，设置不识别：8080 但是 这个却在 bin/pulsar-admin brokers list 和 bin/pulsar-admin   ns-isolation-policy  brokers 出现，这个就会很大的误导用户，我一直以为pulsar认可的broker只是和自己list出来的broker的形式是一样，建议list也不要有端口 或者支持端口形式的ns-isolation-policy

建议2：至少校验primary 的broker 参数是否能匹配到broker 不能匹配应该报错。
校验 如果我设置的--primary  格式不正确或者集群里面匹配不到一台broker 应该报错，如果这不报错 非要等到命中规则报错，那命中规则是我创建了一个topic 这个topic无法分配 然后 cannot lookup 这个就如果是一个人配置的还好 多人配置那查起来就蒙圈了。

### PR DESCRIPTION
**Pull request by tisonkun**
This closes #15795.

### Motivation

`ns-isolation-policy` primary and secondary regex pattern currently verify against hostname only, while we return broker list in hostport pattern, we may support the regex pattern test against the whole URL (or a combine of `host:port` but that can be a bit tricky.

The direction is still to be determinated IMO.

### Modifications

AS WHAT MENTIONED IN MOTIVATION

### Verifying this change

New tests added.

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository:

### CODE REVIEW

File: pulsar-common/src/main/java/org/apache/pulsar/common/policies/impl/NamespaceIsolationPolicyImpl.java [hunk 1]
```diff
@@ -35,24 +37,26 @@
  */
 public class NamespaceIsolationPolicyImpl implements NamespaceIsolationPolicy {
 
-    private List<String> namespaces;
-    private List<String> primary;
-    private List<String> secondary;
-    private AutoFailoverPolicy autoFailoverPolicy;
+    private final List<String> namespaces;
+    private final List<String> primary;
+    private final List<Pattern> primaryPattens;
```
Comment (Technoboy-) [inline pulsar-common/src/main/java/org/apache/pulsar/common/policies/impl/NamespaceIsolationPolicyImpl.java:42]: `primaryPattens` -> `primaryPatterns`

Comment (tisonkun) [reply pulsar-common/src/main/java/org/apache/pulsar/common/policies/impl/NamespaceIsolationPolicyImpl.java]: Fixed 248f31de499acc7deb08801c08e3debb5ac1ea81.

**Review by eolivelli — COMMENTED** [review]
This is a great feature!

@nicoloboschi @michaeljmarshall @lhotari PTAL

I am +1 to it, I am not formally approving it as I believe that this is a new feature and we need to advertise and discuss on the mailing list.

**Comment by tisonkun** [comment]
> advertise and discuss on the mailing list

Let me send one tomorrow :D

**Comment by tisonkun** [comment]
No longer work on this.

**Comment by visxu** [comment]
I have a question to ask, how can I see in which version the problem was fixed ?

**Comment by visxu** [comment]
> This is a great feature!
> 
> @nicoloboschi @michaeljmarshall @lhotari PTAL
> 
> I am +1 to it, I am not formally approving it as I believe that this is a new feature and we need to advertise and discuss on the mailing list.

@eolivelli So, the great feature is still not released, right ? Which version will it plan to be released ?

**Pull request closed without merging**